src/ccl_correlation.c

```diff
@@ -0,0 +1,511 @@
+				   int n_theta,double *theta,double *wtheta,
+				   int corr_type,int *status)
+{
+  corr_int_par *cp=malloc(sizeof(corr_int_par));
+  if(cp==NULL) {
+    *status=CCL_ERROR_MEMORY;
+    strcpy(cosmo->status_message,"ccl_correlation.c: ccl_tracer_corr_bessel ran out of memory\n");
+    return;
+  }
+
+  cp->nell=n_ell;
+  cp->ell0=ell[0];
+  cp->ellf=ell[n_ell-1];
+  cp->cl0=cls[0];
+  cp->clf=cls[n_ell-1];
+  cp->cl_spl=ccl_spline_init(n_ell,ell,cls,cls[0],0);
+  if(cp->cl_spl==NULL) {
+    free(cp);
+    *status=CCL_ERROR_MEMORY;
+    strcpy(cosmo->status_message,"ccl_correlation.c: ccl_tracer_corr_bessel ran out of memory\n");
+    return;
+  }
+  if(corr_type==CCL_CORR_GG)
+    cp->i_bessel=0;
+  else if(corr_type==CCL_CORR_GL)
+    cp->i_bessel=2;
+  else if(corr_type==CCL_CORR_LP)
+    cp->i_bessel=0;
+  else if(corr_type==CCL_CORR_LM)
+    cp->i_bessel=4;
+
+  if(cls[0]*cls[1]<=0)
+    cp->extrapol_0=0;
+  else {
+    cp->extrapol_0=1;
+    cp->tilt0=log10(cls[1]/cls[0])/log10(ell[1]/ell[0]);
+  }
+
+  if(cls[n_ell-2]*cls[n_ell-1]<=0)
+    cp->extrapol_f=0;
+  else {
+    cp->extrapol_f=1;
+    cp->tiltf=log10(cls[n_ell-1]/cls[n_ell-2])/log10(ell[n_ell-1]/ell[n_ell-2]);
+  }
+
+  int ith, gslstatus;
+  double result,eresult;
+  gsl_function F;
+  gsl_integration_workspace *w=gsl_integration_workspace_alloc(ccl_gsl->N_ITERATION);
+  for(ith=0;ith<n_theta;ith++) {
+    cp->th=theta[ith]*M_PI/180;
+    F.function=&corr_bessel_integrand;
+    F.params=cp;
+    gslstatus = gsl_integration_qag(&F, 0, ELL_MAX_FFTLOG, 0,
+                                    ccl_gsl->INTEGRATION_EPSREL, ccl_gsl->N_ITERATION,
+                                    ccl_gsl->INTEGRATION_GAUSS_KRONROD_POINTS,
+                                    w, &result, &eresult);
+    if(gslstatus != GSL_SUCCESS) {
+      ccl_raise_gsl_warning(gslstatus, "ccl_correlation.c: ccl_tracer_corr_bessel():");
+      *status |= gslstatus;
+    }
+    wtheta[ith]=result/(2*M_PI);
+  }
+  gsl_integration_workspace_free(w);
+  ccl_spline_free(cp->cl_spl);
+  free(cp);
+}
+
+
+/*--------ROUTINE: ccl_compute_legendre_polynomial ------
+TASK: Compute input factor for ccl_tracer_corr_legendre
+INPUT: tracer 1, tracer 2, i_bessel, theta array, n_theta, L_max, output Pl_theta
+ */
+static void ccl_compute_legendre_polynomial(int corr_type,int n_theta,double *theta,
+					    int ell_max,double **Pl_theta)
+{
+  int i,j;
+  double k=0;
+
+  //Initialize Pl_theta
+  for (i=0;i<n_theta;i++) {
+    for (j=0;j<ell_max;j++)
+      Pl_theta[i][j]=0.;
+  }
+
+  if(corr_type==CCL_CORR_GG) {
+    for (i=0;i<n_theta;i++) {
+      gsl_sf_legendre_Pl_array(ell_max,cos(theta[i]*M_PI/180),Pl_theta[i]);
+      for (j=0;j<=ell_max;j++)
+	Pl_theta[i][j]*=(2*j+1);
+    }
+  }
+  else if(corr_type==CCL_CORR_GL) {
+    for (i=0;i<n_theta;i++) {//https://arxiv.org/pdf/1007.4809.pdf
+      for (j=2;j<=ell_max;j++) {
+	Pl_theta[i][j]=gsl_sf_legendre_Plm(j,2,cos(theta[i]*M_PI/180));
```

Comment:
These functions do not exist for a fixed m (other than m=0). The only ones that exist dump all ells<ell_max AND all m between 0 and ell for each ell. This is probably overkill for as (especially since the brute-force calculation implemented here is mostly for testing purposes).